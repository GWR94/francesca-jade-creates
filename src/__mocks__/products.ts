export const cakeProduct = {
  id: "8c01cf0d-9de5-4168-8143-73e1ea067ac9",
  title: "Racing Cake",
  description: "Test test test test test test",
  image: [
    {
      bucket: "fjadecreatesmedia121747-production",
      region: "eu-west-2",
      key:
        "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584357819265-racecarcrop.jpg",
    },
    {
      bucket: "fjadecreatesmedia121747-production",
      region: "eu-west-2",
      key:
        "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584357839998-IMG_3565.JPEG",
    },
    {
      bucket: "fjadecreatesmedia121747-production",
      region: "eu-west-2",
      key:
        "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584357864535-IMG_3659.JPG",
    },
  ],
  price: 0,
  shippingCost: 0,
  type: "Cake",
  tags: ["Cake", "Racing"],
  createdAt: "2020-03-09T11:54:02.873Z",
  updatedAt: "2020-03-16T11:24:33.081Z",
  setPrice: false,
};

export const createsProduct = {
  id: "3e743155-95e2-4ae4-8d70-eb5cb2d0c086",
  title: "Name Frame",
  description: "test test test test test test test test test test",
  image: [
    {
      bucket: "fjadecreatesmedia121747-production",
      region: "eu-west-2",
      key:
        "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106241635-IMG_0299.JPEG",
    },
    {
      bucket: "fjadecreatesmedia121747-production",
      region: "eu-west-2",
      key:
        "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106241635-IMG_0299.JPEG",
    },
    {
      bucket: "fjadecreatesmedia121747-production",
      region: "eu-west-2",
      key:
        "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106268144-IMG_0305.JPEG",
    },
  ],
  price: 14.5,
  shippingCost: 4.5,
  type: "Creates",
  tags: ["Personalised", "Name", "Frame"],
  createdAt: "2020-03-09T12:51:52.637Z",
  updatedAt: "2020-03-20T14:24:14.718Z",
  setPrice: true,
};
