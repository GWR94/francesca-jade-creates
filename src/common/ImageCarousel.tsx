import React from "react";
import { Carousel, CarouselIndicators, CarouselItem, CarouselControl } from "reactstrap";
import { S3Image } from "aws-amplify-react";
import { API, Storage, graphqlOperation } from "aws-amplify";
import { Button, Dialog } from "@blueprintjs/core";
import { S3ImageProps } from "./interfaces/Product.i";
import { updateProduct } from "../graphql/mutations";
import { Toaster } from "../utils";

interface Props {
  images: S3ImageProps[];
  deleteImages?: boolean;
  id?: string;
  handleUpdateImages?: (image: S3ImageProps[]) => void;
  update?: boolean;
}

interface State {
  animating: boolean;
  currentIndex: number;
  dialogOpen: boolean;
  keyToDelete: string;
}

class ImageCarousel extends React.Component<Props, State> {
  public readonly state: State = {
    animating: false,
    currentIndex: 0,
    dialogOpen: false,
    keyToDelete: null,
  };

  public handleNextImage = (): void => {
    const { animating, currentIndex } = this.state;
    const { images } = this.props;
    if (animating) return;
    this.setState({
      currentIndex: currentIndex === images.length - 1 ? 0 : currentIndex + 1,
    });
  };

  public handlePreviousImage = (): void => {
    const { animating, currentIndex } = this.state;
    const { images } = this.props;
    if (animating) return;
    this.setState({
      currentIndex: currentIndex === 0 ? images.length - 1 : currentIndex - 1,
    });
  };

  public goToIndex = (idx: number): void => {
    const { animating } = this.state;
    if (animating) return;
    this.setState({ currentIndex: idx });
  };

  public handleDeleteImage = async (): Promise<void> => {
    const { keyToDelete } = this.state;
    const { images, update, id, handleUpdateImages } = this.props;
    try {
      const updatedImages = images.filter((img) => img.key !== keyToDelete);
      this.setState({ currentIndex: 0 });
      await Storage.remove(keyToDelete);
      if (update) {
        await API.graphql(
          graphqlOperation(updateProduct, { input: { id, image: updatedImages } }),
        );
      }
      Toaster.show({
        intent: "success",
        message: "Successfully removed image.",
      });
      handleUpdateImages(updatedImages);
      // setDialogOpen(false);
      this.setState({ animating: false });
    } catch (err) {
      console.error(err);
      Toaster.show({
        intent: "danger",
        message: "Failed to removed image. Please try again.",
      });
    }
  };

  public render(): JSX.Element {
    const { images, deleteImages } = this.props;
    const { currentIndex, dialogOpen, keyToDelete } = this.state;
    return (
      <>
        <div className="carousel__container">
          {images.length === 1 ? (
            <div className="update__image-container">
              {deleteImages && (
                <i
                  className="fas fa-times carousel__delete-icon animated infinite pulse"
                  role="button"
                  tabIndex={0}
                  onClick={(): void => {
                    this.setState({
                      dialogOpen: false,
                      keyToDelete: images[0].key,
                    });
                  }}
                />
              )}
              <S3Image
                imgKey={images[0].key}
                theme={{
                  photoImg: {
                    width: "100%",
                  },
                }}
              />
            </div>
          ) : (
            <Carousel
              activeIndex={currentIndex}
              next={this.handleNextImage}
              previous={this.handlePreviousImage}
            >
              <CarouselIndicators
                items={images}
                activeIndex={currentIndex}
                onClickHandler={(idx): void => this.setState({ currentIndex: idx })}
                className="carousel__indicators"
              />
              {images.map((image, i) => (
                <CarouselItem
                  onExiting={(): void => this.setState({ animating: true })}
                  onExited={(): void => this.setState({ animating: false })}
                  key={i}
                >
                  {deleteImages && (
                    <i
                      className="fas fa-times carousel__delete-icon animated infinite pulse"
                      role="button"
                      tabIndex={0}
                      onClick={(): void => {
                        this.setState({
                          dialogOpen: true,
                          keyToDelete: image.key,
                        });
                      }}
                    />
                  )}
                  <div className="update__image-container">
                    <S3Image
                      imgKey={image.key}
                      theme={{
                        photoImg: {
                          width: "100%",
                        },
                      }}
                    />
                  </div>
                </CarouselItem>
              ))}
              <CarouselControl
                direction="prev"
                directionText="Previous"
                onClickHandler={this.handlePreviousImage}
              />
              <CarouselControl
                direction="next"
                directionText="Next"
                onClickHandler={this.handleNextImage}
              />
            </Carousel>
          )}
        </div>
        <Dialog
          isOpen={dialogOpen}
          icon="trash"
          onClose={(): void => this.setState({ dialogOpen: false })}
          title="Are you sure?"
        >
          <div className="update__alert-container">
            <p className="text-center">Are you sure you want to delete this image?</p>
            <p className="text-center">Other images can be added at a later date.</p>

            <S3Image
              imgKey={keyToDelete}
              theme={{
                photoImg: { maxWidth: "100%", marginBottom: "20px" },
              }}
            />
            <div className="dialog__button-container">
              <Button
                intent="danger"
                text="Cancel"
                onClick={(): void => this.setState({ dialogOpen: false })}
                style={{ margin: "0 5px" }}
              />
              <Button
                intent="success"
                text="Confirm"
                onClick={this.handleDeleteImage}
                style={{ margin: "0 5px" }}
              />
            </div>
          </div>
        </Dialog>
      </>
    );
  }
}

export default ImageCarousel;
