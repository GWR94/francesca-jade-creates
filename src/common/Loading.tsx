import React from "react";
import { CircularProgress, makeStyles } from "@material-ui/core";

interface LoadingProps {
  size?: number;
  color?: "inherit" | "primary" | "secondary";
  small?: boolean;
}

/**
 * A component which shows a loading spinner to the user to signal that there is
 * some loading going on.
 * @param {number} size - The size of the loading spinner
 * @param {string} color - The colour of the loading spinner
 */
const Loading: React.FC<LoadingProps> = ({
  size = 100,
  color = "inherit",
  small = false,
}): JSX.Element => {
  const useStyles = makeStyles({
    loading: {
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
      width: "100%",
      height: "calc(100vh - 60px)",
    },
    loadingSmall: {
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
      width: "100%",
      padding: 60,
    },
    spinner: {
      margin: "0 auto",
      display: "block",
    },
  });
  const classes = useStyles();
  return (
    <div className={small ? classes.loadingSmall : classes.loading}>
      <CircularProgress size={size} color={color} className={classes.spinner} />
    </div>
  );
};

export default Loading;
