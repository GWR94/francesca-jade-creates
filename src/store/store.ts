import { createStore, combineReducers, compose, Store } from "redux";
import { persistStore, persistReducer } from "redux-persist";
import storageSession from "redux-persist/lib/storage/session";
import userReducer, { UserState } from "../reducers/user.reducer";
import basketReducer, { BasketState } from "../reducers/basket.reducer";

declare global {
  interface Window {
    // eslint-disable-next-line no-undef
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose;
  }
}

export interface AppState {
  basket: BasketState;
  user: UserState;
}

const persistConfig = {
  key: "root",
  storage: storageSession,
};

const rootReducer = combineReducers({
  basket: basketReducer,
  user: userReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export default (): any => {
  const store: Store<AppState> = createStore(persistedReducer, composeEnhancers());
  const persistor = persistStore(store);
  return { store, persistor };
};
