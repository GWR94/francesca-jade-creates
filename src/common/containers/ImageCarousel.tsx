import React from "react";
import {
  CarouselProvider,
  Slider,
  Slide,
  ButtonBack,
  ButtonNext,
} from "pure-react-carousel";
import "pure-react-carousel/dist/react-carousel.es.css";
import { S3Image } from "aws-amplify-react";
import { API, Storage, graphqlOperation } from "aws-amplify";
import { ChevronLeft, ChevronRight } from "@material-ui/icons";
import { updateProduct } from "../../graphql/mutations";
import { openSnackbar } from "../../utils/Notifier";
import { ImageCarouselProps, ImageCarouselState } from "./interfaces/ImageCarousel.i";
import { COLORS, INTENT } from "../../themes";
import DeleteImageDialog from "../alerts/DeleteImageDialog";

/**
 * Class component for displaying, updating and removing images for a specific product.
 */
class ImageCarousel extends React.Component<ImageCarouselProps, ImageCarouselState> {
  public readonly state: ImageCarouselState = {
    dialogOpen: false,
    keyToDelete: null,
    isPlaying: true,
  };

  // Ref of carousel so event listeners can be mounted/unmounted to it.
  public carouselRef = React.createRef<HTMLDivElement>();

  /**
   * When the component mounts an event listener should be added to the carousel
   * so it can be paused when hovering and play when removing the mouse from the
   * area - but only when there is 1 or more images in the array.
   */
  public componentDidMount(): void {
    const { images } = this.props;
    if (images.length > 1 && this.carouselRef.current) {
      this.carouselRef.current.addEventListener("mouseenter", (): void => {
        this.setState({ isPlaying: false });
      });
      this.carouselRef.current.addEventListener("mouseleave", (): void => {
        this.setState({ isPlaying: true });
      });
    }
  }

  /**
   * Remove the event listeners when the component unmounts to prevent memory
   * leaks.
   */
  public componentWillUnmount(): void {
    const { images } = this.props;
    if (images.length > 1 && this.carouselRef.current) {
      this.carouselRef.current.removeEventListener("mouseenter", (): void => {
        this.setState({ isPlaying: false });
      });
      this.carouselRef.current.removeEventListener("mouseleave", (): void => {
        this.setState({ isPlaying: true });
      });
    }
  }

  /**
   * Method to remove an image from the carousel, which will inturn remove the
   * image from the database and sever the link between the image and the current
   * product.
   */
  public handleDeleteImage = async (): Promise<void> => {
    // destructure the key to delete!bet wi
    const { keyToDelete } = this.state;
    const { images, update, id, handleUpdateImages } = this.props;
    /**
     * if there is no value in keyToDelete, then the operation cannot be completed,
     * so notify the user of this with a danger snackbar, and return from the function.
     */
    if (!keyToDelete) {
      openSnackbar({
        severity: INTENT.Danger,
        message: "Invalid key to delete. Please try again.",
      });
      return;
    }
    try {
      /**
       * filter out the key (which is saved in state from clicking the remove button
       * on the image in the carousel), so that it's been removed from the array, and
       * therefore updated to its correct data in state.
       */
      const updatedImages = images.filter((img) => img.key !== keyToDelete);
      /**
       * remove the image from the database by providing the key to delete to
       * aws-amplify Storage library
       */
      await Storage.remove(keyToDelete);

      /**
       * If the update prop is set to true, then the link from the image to the product
       * has to be severed, which can be done using the updateProduct graphQL mutation.
       * The previously created updatedImages needs to be passed for the image property
       * so the images left are the correct ones.
       */
      if (update) {
        await API.graphql(
          graphqlOperation(updateProduct, { input: { id, image: updatedImages } }),
        );
      }
      /**
       * Once the operation is complete, let the user know with a success snackbar with
       * a relevant message.
       */
      openSnackbar({
        severity: "success",
        message: "Successfully removed image.",
      });
      // if the handleUpdateImages function is defined, execute it.
      if (handleUpdateImages) handleUpdateImages(updatedImages);
      this.setState({ dialogOpen: false });
    } catch (err) {
      /**
       * If there are any errors at any point in the function, notify the user with a
       * danger snackbar with a relevant message.
       */
      openSnackbar({
        severity: "error",
        message: "Failed to removed image. Please try again.",
      });
    }
  };

  public render(): JSX.Element {
    const { images, deleteImages, type } = this.props;
    const { dialogOpen, keyToDelete, isPlaying } = this.state;
    return (
      <>
        {/* 
          If there is only one item in the images array, then there is no need to render
          the carousel component, so just place the image in a div.
        */}
        {images.length === 1 ? (
          <div style={{ position: "relative" }}>
            {deleteImages && (
              /**
               * if the user is allowed to delete images (is an admin), show a pulsing
               * delete icon which shows the user that they can delete it
               */
              <i
                className="fas fa-times carousel__delete-icon animated infinite pulse"
                role="button"
                tabIndex={0}
                onClick={(): void => {
                  this.setState({
                    dialogOpen: true,
                    keyToDelete: images[0].key,
                  });
                }}
              />
            )}
            <S3Image
              imgKey={images[0].key}
              theme={{
                photoImg: {
                  width: "100%",
                },
              }}
            />
          </div>
        ) : (
          /**
           * If there is more than one image in the array then the carousel needs to be
           * rendered with each of the images mapped to a Slide component.
           */
          <div ref={this.carouselRef} style={{ position: "relative" }}>
            <CarouselProvider
              naturalSlideHeight={6}
              naturalSlideWidth={5}
              totalSlides={images.length}
              visibleSlides={1}
              step={1}
              isPlaying={isPlaying && !dialogOpen}
              infinite
            >
              <Slider>
                {images.map((image, i) => (
                  <Slide index={i} key={i} style={{ position: "relative" }}>
                    {deleteImages && (
                      <i
                        className="fas fa-times carousel__delete-icon animated infinite pulse"
                        role="button"
                        tabIndex={0}
                        onClick={(): void => {
                          this.setState({
                            dialogOpen: true,
                            keyToDelete: image.key,
                          });
                        }}
                      />
                    )}
                    <S3Image
                      imgKey={image.key}
                      theme={{
                        photoImg: {
                          width: "100%",
                        },
                      }}
                    />
                  </Slide>
                ))}
              </Slider>
              {/*  Buttons to control the playback of the carousel */}
              <ButtonBack className="carousel__back animated pulse infinite">
                <ChevronLeft
                  className="carousel__svg"
                  style={{ color: type === "Cake" ? COLORS.Pink : COLORS.Purple }}
                />
              </ButtonBack>
              <ButtonNext className="carousel__forward animated pulse infinite">
                <ChevronRight
                  className="carousel__svg"
                  style={{ color: type === "Cake" ? COLORS.Pink : COLORS.Purple }}
                />
              </ButtonNext>
            </CarouselProvider>
          </div>
        )}
        {/* 
          Dialog to confirm if the user wants to delete the image. Only opens when
          the user clicks the delete icon on the image. 
        */}
        {keyToDelete && (
          <DeleteImageDialog
            dialogOpen={dialogOpen}
            closeDialog={(): void => this.setState({ dialogOpen: false })}
            keyToDelete={keyToDelete}
            handleDeleteImage={this.handleDeleteImage}
          />
        )}
      </>
    );
  }
}

export default ImageCarousel;
