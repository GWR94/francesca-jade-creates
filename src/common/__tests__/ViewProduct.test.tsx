import React from "react";
import { shallow } from "enzyme";
import { Tag, Button } from "@blueprintjs/core";
import configureStore from "redux-mock-store";
import ConnectedViewProduct, {
  ViewProduct,
} from "../../pages/accounts/components/ViewProduct";
import Loading from "../Loading";
import { cakeProduct, createsProduct, userAttributes } from "../../__mocks__/products";
import ImageCarousel from "../containers/ImageCarousel";
import { AppState } from "../../store/store";
import * as types from "../../interfaces/basket.redux.i";

describe("<ViewProduct />", () => {
  describe("test cases for unconnected ViewProduct component", () => {
    it("should render the loading component when there is no product", () => {
      const wrapper = shallow<ViewProduct>(
        <ViewProduct id="123" userAttributes={userAttributes} />,
      );
      expect(wrapper.find(Loading).length).toBe(1);
      expect(wrapper).toMatchSnapshot();
    });

    it("should render correctly when injecting product with no set price into state", () => {
      const wrapper = shallow<ViewProduct>(
        <ViewProduct id={cakeProduct.id} userAttributes={userAttributes} />,
      );
      // inject product into state
      wrapper.setState({ product: cakeProduct });
      expect(wrapper).toMatchSnapshot();
      // expect title to be correct based on product
      expect(wrapper.find("h3.view__title").text()).toBe("Racing Cake");
      // expect description to be correct based on product
      expect(wrapper.find(".view__description").text()).toBe(
        "Test test test test test test",
      );
      // expect all children to be of type tag and have the correct text in them.
      expect(wrapper.find(".view__tags").childAt(0).type()).toBe(Tag);
      expect(wrapper.find(Tag).at(0).childAt(0).text()).toBe(cakeProduct.tags[0]);
      expect(wrapper.find(".view__tags").childAt(1).type()).toBe(Tag);
      expect(wrapper.find(Tag).at(1).childAt(0).text()).toBe(cakeProduct.tags[1]);
      // expect image carousel prop "images" to match the products image prop
      expect(wrapper.find(ImageCarousel).prop("images")).toBe(cakeProduct.image);
      // expect price to show correct text based on product
      expect(wrapper.find(".view__price").childAt(0).text()).toBe(
        "Please send a request with your personalisation preferences, and I will get back to you as soon as possible with a quote.",
      );
      // expect there to only be one button when there is no set price
      expect(wrapper.find(Button).length).toBe(1);
      // expect the button to have "success" intent prop
      expect(wrapper.find(Button).prop("intent")).toBe("success");
      // expect the icon prop to be correct based on the product
      expect(wrapper.find(Button).prop("icon")).toStrictEqual(
        <i className="fas fa-credit-card view__icon" />,
      );
      // expect second child of button to be the correct text
      expect(wrapper.find(Button).prop("text")).toBe("Request a Quote");
    });

    it("should render correctly when injecting product with set price into state", () => {
      const wrapper = shallow<ViewProduct>(
        <ViewProduct id={createsProduct.id} userAttributes={userAttributes} />,
      );
      // inject product into state
      wrapper.setState({ product: createsProduct });
      expect(wrapper).toMatchSnapshot();
      // expect title to be correct based on product
      expect(wrapper.find("h3.view__title").text()).toBe(createsProduct.title);
      // expect description to be correct based on product
      expect(wrapper.find(".view__description").text()).toBe(createsProduct.description);
      // expect all children to be of type tag and have the correct text in them.
      expect(wrapper.find(".view__tags").childAt(0).type()).toBe(Tag);
      expect(wrapper.find(Tag).at(0).childAt(0).text()).toBe(createsProduct.tags[0]);
      expect(wrapper.find(".view__tags").childAt(1).type()).toBe(Tag);
      expect(wrapper.find(Tag).at(1).childAt(0).text()).toBe(createsProduct.tags[1]);
      // expect image carousel prop "images" to match the products image prop
      expect(wrapper.find(ImageCarousel).prop("images")).toBe(createsProduct.image);
      // expect price to show correct text based on product price
      expect(wrapper.find(".view__price").childAt(0).text()).toBe(
        `The cost for ${createsProduct.title} is £${createsProduct.price.toFixed(
          2,
        )} + £${createsProduct.shippingCost.toFixed(2)} postage and packaging`,
      );
      // expect there to only be one button
      expect(wrapper.find(Button).length).toBe(1);
      // expect the button to have "success" intent prop
      expect(wrapper.find(Button).prop("intent")).toBe("primary");
      // expect the icon prop to be correct based on the prop
      expect(wrapper.find(Button).prop("icon")).toStrictEqual(
        <i className="fas fa-shopping-cart view__icon" />,
      );
      // expect text of button to be correct for adding to basket
      expect(wrapper.find(Button).prop("text")).toBe("Add to Basket");
    });
  });

  describe("test cases for connected ViewProduct component", () => {
    let wrapper;
    // mock store to test actions are created/dispatched.
    const mockStore = configureStore();
    let store;
    const initialState: AppState = {
      basket: {
        items: [],
        cost: 0,
      },
      user: {
        id: null,
      },
    };
    // before each test, reset store to initial state and clear actions.
    beforeEach(() => {
      store = mockStore(initialState);
      wrapper = shallow(
        <ConnectedViewProduct id="123" userAttributes={userAttributes} store={store} />,
      );
    });

    it("should render correctly", () => {
      expect(wrapper.find(ViewProduct).length).toBe(1);
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.prop("store").getState()).toBe(initialState);
    });

    it("should dispatch actions to mock store and update state", () => {
      wrapper = wrapper.dive();
      wrapper.setState({ product: createsProduct });
      expect(wrapper).toMatchSnapshot();
      wrapper.find(Button).simulate("click");
      const { id, title, description, price, shippingCost, image, type } = createsProduct;
      expect(store.getActions()[0]).toStrictEqual({
        type: types.ADD_ITEM_TO_BASKET,
        item: {
          id,
          title,
          description,
          price,
          shippingCost,
          image,
          type,
        },
      });
    });
  });
});
