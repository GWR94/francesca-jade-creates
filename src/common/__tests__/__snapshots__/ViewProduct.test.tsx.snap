// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ViewProduct /> should render correctly when rendering a product of type 'Cake' 1`] = `<Loading />`;

exports[`<ViewProduct /> should render the loading component when there is no product 1`] = `<Loading />`;
